### DISCUSSION TITLE
How to install Github Copilot on code-server?

### DISCUSSION THREAD

**Discussion by yuchen1204** [coder/code-server#5063] [Q&A] [answered]
![Screenshot 2022-04-04 06 30 05](https://user-images.githubusercontent.com/92127505/161451517-f8eb650e-fefc-4dd0-8bbc-222f77264548.png)

**Comment by lg-nineoneninethreenine** [comment, marked as answer]
Firstly, update your code-server instance and then retry. I seem to have noticed that you  use Code 1.50.0, not the latest 1.64.*, hence run the install command again and launch code-server.

**Reply by yuchen1204** [reply]
I successfully installed github capilot on the code server, thank you

**Reply by lg-nineoneninethreenine** [reply]
😀welcome

**Reply by gukuhn** [reply]
hello! I have the latest code-server installed (1.66.2) but it shows the same issue (Unable to install extension 'github.copilot' as it is not compatible with VS Code '1.66.2'.). Could you help me? Thanks!!

**Reply by lg-nineoneninethreenine** [reply]
Can you share a screenshot or video of the same and which guide you followed?

**Reply by gukuhn** [reply]
I tried install a older version of the VSIX file and it worked!

thanks anyways!

**Reply by yuchen1204** [reply]
github capilot can't use on code-server
![image](https://user-images.githubusercontent.com/92127505/173103781-95f1b99c-3c19-4919-af5f-114c15bc9881.png)

**Reply by yuchen1204** [reply]
![image](https://user-images.githubusercontent.com/92127505/173104493-320b23d2-2fae-4365-974d-847f967dc87f.png)

**Reply by jsjoeio** [reply]
We're working on a VS Code upgrade soon!

**Comment by elgraniti** [comment]
Any updates on how to install copilot on code server 4.7.0?

**Reply by TheMcSebi** [reply]
Download the vsix file from the marketplace.visualstudio.com and install it on the command palette (Ctrl+Shift+P, "vsix", Enter)

**Reply by xKliment** [reply]
I get this error for some reason "Command 'Extensions: Install from VSIX...' resulted in an error"

**Reply by ShubhamC1209** [reply]
Same i got the same error- Command 'Extensions: Install from VSIX...' resulted in an error extension/package.json not found inside zip.

**Reply by Neo11Neo** [reply]
works if you download the visx from their marketplace: [marketplace.visualstudio.com](https://marketplace.visualstudio.com/items?itemName=GitHub.copilot)

**Reply by Nagico** [reply]
And I need to decompress the vsix file before install, see here: https://stackoverflow.com/questions/66134532/vscode-marketplace-extension-corrupt-zip-end-of-central-directory-record-signa

**Comment by Alababdiy** [comment]
I found this video on YouTube that worked for me: https://www.youtube.com/watch?v=w_-zFFM0N-g. It may help you install Github Copilot on code-server as well.

**Comment by josephrocca** [comment]
## August 2023

I just ran into this problem (Copilot doesn't show up when you search for it in extensions), so here's what worked for me:

1. Visit https://marketplace.visualstudio.com/items?itemName=GitHub.copilot
2. Look for a "Download Extension" link. As of writing it's in the side-bar under the "Resources" header.
3. You'll get a file called something like `GitHub.copilot-1.100.313.vsix`. Drag and drop it anywhere in your filetree in code-server.
4. Right click on it, and you'll see an option "Install Extension VSIX" - click it.
5. It'll install, and then after a moment or two you'll see a notification in the bottom-right, prompting you to "Sign in to Github to use Copilot". Click sign in, and follow the prompts.
6. That's it - test it by typing some code. You should see the grey text auto-complete come up.
 
Note that before doing the above steps I upgraded to the latest version of `code-server` (VSCode=v1.80.2, code-server=v4.16.1) and restarted the machine.

**Reply by staspiter** [reply]
It works very well. However, when I refresh my code-server page, it loses Github authentication and the "Sign in to Github" prompt shows up again.

**Reply by josephrocca** [reply]
Yep, I have the same problem. If anyone finds a fix, please do share!

**Reply by ashokkumar14440** [reply]
I had the same issue so I downgraded from [v4.16.1](https://github.com/coder/code-server/releases/tag/v4.16.1) to [v4.16.0](https://github.com/coder/code-server/releases/tag/v4.16.0). After downgrade the git credentials are retained in memory even after a reload.

**Reply by josephrocca** [reply]
@ashokkumar14440 This didn't work for me - does it retain after reboot for you? Steps I took:

1. Install 4.16.0: `curl -fsSL https://code-server.dev/install.sh | sh -s -- --version 4.16.0`
2. Confirm that `code-server --version` matches `4.16.0`'s `code` version (`1.80.1`)
3. Reboot
4. Sign into the Github when prompted
5. Reboot
6. I am asked to sign in to Github again

**Reply by bravedave** [reply]
#5052 It's a bug - tragic - I don't think the downgrade helps, it's been there for a while

Thanks for the how to but, helps heaps

**Reply by ashokkumar14440** [reply]
@josephrocca If I shut down and restart my SageMaker server credentials are not retained, In the latest version even reloading the CodeServer webpage on Sagemaker(F5 Key) loses the git credentials, downgrade helped with the reload issue but not the restart issue.

**Reply by mynameisnewinjp** [reply]
Indeed, while GitHub connections are disconnected upon reboot, Copilot can be used without logging into GitHub each time.

The chat feature is not as usable as it is in Codespaces.

**Reply by code-asher** [reply]
#5052 should have been fixed in v4.17.1 (see https://github.com/coder/code-server/issues/5072) but if anyone is still seeing issues please do open an issue with steps to reproduce.

**Reply by giovp** [reply]
confirming it worked for me with 4.20.1 with Code 1.85.2

**Reply by Venkythota1** [reply]
I tried the above approach, when I try to sign in into the GitHub copilot. I get this pop up of **You have not yet finished authorizing this extension to use GitHub. Would you like to try a different way? (personal access token) Source: GitHub Authentication** when i click on yes. it asks to create a personal acess token when I get give the personal token, I am getting this below error in the output window.

auth: Error: No auth flow succeeded.
at a.GitHubServer login (/code-server-4.93.1-linux-amd64/lib/vscode/extensions/github-authentication/dist/extension js 1:228535)
at process TicksAndRejections (node internal/process/task_queues:95:5)
at a.GitHubAuthenticationProvider.createSession (/code-server-4.93.1-linux-amd64/lib/vscode/extensions/github- authentication/dist/extension.js:1:224358)
at c. $createSession (/code-server-4.93.1-linux-amd64/lib/vscode/out/vs/workbench/api/node/out-vscode-reh-
web/vs/workberich/api/node/file:/home/runner/work/code-server/code- server/lib/vscode/src/vs/workbench/api/common/extHostAuthentication.ts:89:11)

I am using the latest github copilot and vs-codeserver (v4.93.1) vs code (1.93.1)

is there any workaround for this issue?

**Comment by hchaudhary1** [comment]
I downloaded the vsix chat extension, unzipped it and modified the `package.json` to set the following:   

```
"vscode": "^1.84.0",
```

rezipped it and did a manual vsix installed, and it now works on latest code server.

**Comment by ksylvan** [comment]
I installed GitHub Copilot (the latest version v1.168.742) with code-server v4.21.2 (Code 1.86.2).

I wasn't able to get GitHub Copilot Chat working. Not even trying to use @hchaudhary1 method above, changing the json file to have:

```
"vscode": "^1.86.0".
```

And after a successful install, this happened:

![image](https://github.com/coder/code-server/assets/1226059/c44d3654-6800-48f8-a487-3af5b7c782b1)

**Comment by eliluong** [comment]
Hello everyone, I wanted to offer another solution I just ran across, because I had a similar issue.

I am running code-server 4.22.1 inside Docker (code-server: v4.22.1, Code: 1.87.2, Commit: 760d1318e98945d05133c6121f99541ca7a39bf8). I downloaded the extension from [here](https://marketplace.visualstudio.com/items?itemName=GitHub.copilot), and installed inside code-server (uploaded, right-click to install), but the authentication is always failing.

My solution was to start the sign-in process. After it gives you the 8-character code, copy it somewhere. Then when it prompts you to go to the URL to activate, **do not click Open** but instead Copy the URL (it should be https://github.com/login/device anyway). Then open a new tab yourself, paste the 8-character code, Authorize it, then close the tab.

And then it should successfully log in :)

**Reply by jamesbiederbeck** [reply]
If you want to automate the extension installation, you can use https://marketplace.visualstudio.com/_apis/public/gallery/publishers/GitHub/vsextensions/copilot/<version_id>/vspackage

1.206.0 was the version I found at time of writing.

**Comment by jamesy0ung** [comment]
If you are running in the Linuxserver code-server container, add this to your environment
      - "EXTENSIONS_GALLERY={\"serviceUrl\": \"https://marketplace.visualstudio.com/_apis/public/gallery\", \"cacheUrl\": \"https://vscode.blob.core.windows.net/gallery/index\", \"itemUrl\": \"https://marketplace.visualstudio.com/items\"}"

**Reply by nijuyonkadesu** [reply]
https://coder.com/docs/code-server/FAQ#why-cant-code-server-use-microsofts-extension-marketplace

your solution conflicts with this FAQ section I believe

**Comment by Venkythota1** [comment]
Hi everyone, 
I tried the above approach, when I try to sign in into the GitHub copilot. I get this pop up of You have not yet finished authorizing this extension to use GitHub. Would you like to try a different way? (personal access token) Source: GitHub Authentication when i click on yes. it asks to create a personal acess token when I get give the personal token, I am getting this below error in the output window.

auth: Error: No auth flow succeeded.
at a.GitHubServer login (/code-server-4.93.1-linux-amd64/lib/vscode/extensions/github-authentication/dist/extension js 1:228535)
at process TicksAndRejections (node internal/process/task_queues:95:5)
at a.GitHubAuthenticationProvider.createSession (/code-server-4.93.1-linux-amd64/lib/vscode/extensions/github- authentication/dist/extension.js:1:224358)
at c. $createSession (/code-server-4.93.1-linux-amd64/lib/vscode/out/vs/workbench/api/node/out-vscode-reh-
web/vs/workberich/api/node/file:/home/runner/work/code-server/code- server/lib/vscode/src/vs/workbench/api/common/extHostAuthentication.ts:89:11)

I am using the latest github copilot and vs-codeserver (v4.93.1) vs code (1.93.1)

is there any workaround for this issue?

**Comment by robvanoostenrijk** [comment]
Grab latest copilot VSIX package automatically:

```
#!/bin/bash

VERSION=$(curl -s 'https://marketplace.visualstudio.com/_apis/public/gallery/extensionquery' \
  -H 'accept: application/json;api-version=7.2-preview.1;excludeUrls=true' \
  --json '{"filters":[{"criteria":[{"filterType":7,"value":"github.copilot"}]}],"flags":2151}' | jq -r '.results[0].extensions[0].versions[0].version')

echo "[i] Version ${VERSION}"

curl \
        -o "copilot-${VERSION}.vsix" \
        --compressed \
        "https://marketplace.visualstudio.com/_apis/public/gallery/publishers/GitHub/vsextensions/copilot/${VERSION}/vspackage"
```

**Comment by ANGJustinl** [comment]
![image](https://github.com/user-attachments/assets/39e8ddd5-5ec3-4729-9e4f-5fb5c1e7cab1)
`Unable to install extension 'github.copilot-chat' as it is not compatible with VS Code '1.100.2'.`

Using code-server: v4.100.2
version: 1.100.2
commit: 9dd999ba78032163d606fadd531de5e1fcae5443
data: 2025-05-15T19:04:25.944Z (15 h ago)

Installing `GitHub.copilot-chat-0.28.2025051601.vsix`

**Reply by v0lkv** [reply]
Yes, I also expect compatibility of these versions. It seems copilot works with 101+. The code server is taking too long to update(

**Reply by code-asher** [reply]
1.100.2 is the latest stable version of VS Code.  It looks like you may be installing a version of the extension for VS Code insiders.  We only release for the stable release.

**Reply by Nero3ooo** [reply]
Same problem here, downloaded from:
https://marketplace.visualstudio.com/items?itemName=GitHub.copilot-chat
tested with copilot-chat versions 0.28.2025052001 and 0.28.2025051601
I could not find an older version for this extension

**Reply by code-asher** [reply]
I think the only way to get an old enough version might be to use their API directly.  But, looks like things might change here soon anyway: https://code.visualstudio.com/blogs/2025/05/19/openSourceAIEditor

> We will open source the code in the [GitHub Copilot Chat extension](https://marketplace.visualstudio.com/items?itemName=GitHub.copilot-chat) under the MIT license

**Reply by webleedev** [reply]
version ~~`0.27.2025050603`~~ `0.26.2025040204` worked for me with 1.100.3. It seems compatibility breaks with the introduction of agent mode

**Reply by Kahdeg-15520487** [reply]
i managed to install version 0.27.2025050603, how do i open the chat windows

**Reply by Kahdeg-15520487** [reply]
i found the hotkey to toggle the chat windows, but got hit with this
![image](https://github.com/user-attachments/assets/d9a35645-f682-4d93-add2-18b86c534197)

**Reply by rasheed-fabrichq** [reply]
Same here. used  version 0.27.2025050603 and getting the above options. Is it because the  version 0.27.2025050603 is not there in the versions listed in [here](https://marketplace.visualstudio.com/items?itemName=GitHub.copilot-chat&ssr=false#version-history)?

@Kahdeg-15520487 , Please let me know if you find any solution for this

**Reply by webleedev** [reply]
Thanks for highlighting this @Kahdeg-15520487 . I think my chat panel somehow cached an old view which led me to believe it was working. I've updated my previous comment

**Reply by rasheed-fabrichq** [reply]
Hi @webleedev , I was able to install github-copilot-chat version 0.26.2025040204. However, it doesn’t seem to be interactive — I’m unable to chat with it, and there’s no response when I try. Is it working as expected for you on your end?
Sharing a screenshot below for reference.

![Screenshot from 2025-06-18 19-02-56](https://github.com/user-attachments/assets/c198527a-c58d-49f3-9dff-33450563cda0)

**Reply by arttkach** [reply]
Installed copilot with:

- `curl -o "copilot-chat-0.26.2025040204.vsix" --compressed "https://marketplace.visualstudio.com/_apis/public/gallery/publishers/GitHub/vsextensions/copilot-chat/0.26.2025040204/vspackage"`
- `code-server --install-extension copilot-chat-0.26.2025040204.vsix`

code-server: `4.102.2`
version: `1.102.2`

Chat doesn't work. Switching to the release version doesn't work either:

<img width="502" height="771" alt="image" src="https://github.com/user-attachments/assets/dd4a4e4f-496b-429d-837e-b0a726f5fed8" />

**Reply by Chris-LewisI** [reply]
I just tried your same installation method and I don't get the popup that asks me to sign in.

**Reply by arttkach** [reply]
IIRC the sign in happened for me when I installed GitHub copilot in addition to chat.

**Comment by RodrigoRP** [comment]
Has anyone managed to install both GitHub Copilot extensions on the latest code-server version 4.1.03.2?

**Comment by oxFilla** [comment]
This works for me:

```
mkdir -p /tmp/extensions && \
curl -L "https://marketplace.visualstudio.com/_apis/public/gallery/publishers/GitHub/vsextensions/copilot/latest/vspackage" \
-o /tmp/extensions/github-copilot.vsix && \
gunzip -c /tmp/extensions/github-copilot.vsix > /tmp/extensions/github-copilot-decompressed.vsix && \
/usr/local/bin/code-server --install-extension /tmp/extensions/github-copilot-decompressed.vsix && \
rm -rf /tmp/extensions
```

**Comment by lennardv2** [comment]
I created this https://github.com/sunpix/howto-install-copilot-in-code-server

installs latest compatible:
- copilot
- copilot-chat

**Reply by johny-mnemonic** [reply]
Thanks for this script!
I had to modify it as the version output doesn't match it's expectation.
To get the version number I switched to `code-server --version | head -n 1` as the line contains only the number, no "with Code" text is there. Maybe it's because I am using code-server delivered by Home Assistant, which is slightly outdated (1.99.3).

**Reply by lennardv2** [reply]
In the meantime i've gone the route of using the official vscode, with the --serve-web option. This way it's always up to date with an official marketplace, and copilot is just installable from there.

See for example:
https://github.com/nerasse/my-code-server

One disadvantage i noticed: there doesnt seem to be a way activate the portforwarding with the --serve-web version

**Reply by lennardv2** [reply]
> Maybe it's because I am using code-server delivered by Home Assistant, which is slightly outdated (1.99.3).

Yes that's probably it

**Reply by goodmanvn2003** [reply]
Thanks @johny-mnemonic 
I put the script for y'all convenience
```shell
# Extract VS Code version from code-server
get_vscode_version() {
    code-server --version 2>/dev/null | head -n 1
}

# Get user-data-dir from running code-server process
get_user_data_dir() {
    # Use ps with POSIX-compliant options
    local process_info
    if command -v ps >/dev/null 2>&1; then
        # Try BSD-style first (macOS), fallback to POSIX
        process_info=$(ps aux 2>/dev/null | grep -v grep | grep "code-server" | head -n 1) ||
        process_info=$(ps -ef 2>/dev/null | grep -v grep | grep "code-server" | head -n 1)
    fi

    if [ -n "$process_info" ]; then
        echo "$process_info" | grep -o -- '--user-data-dir=[^ ]*' | sed 's/--user-data-dir=//'
    fi
}

# Find compatible extension version
find_compatible_version() {
    local extension_id="$1"
    local vscode_version="$2"

    local response
    response=$(curl -s -X POST "https://marketplace.visualstudio.com/_apis/public/gallery/extensionquery" \
        -H "Content-Type: application/json" \
        -H "Accept: application/json;api-version=3.0-preview.1" \
        -d "{
            \"filters\": [{
                \"criteria\": [
                    {\"filterType\": 7, \"value\": \"$extension_id\"},
                    {\"filterType\": 12, \"value\": \"4096\"}
                ],
                \"pageSize\": 50
            }],
            \"flags\": 4112
        }")

    echo "$response" | jq -r --arg vscode_version "$vscode_version" '
        .results[0].extensions[0].versions[] |
        select(.version | test("^[0-9]+\\.[0-9]+\\.[0-9]*$")) |
        select(.version | length < 8) |
        {
            version: .version,
            engine: (.properties[] | select(.key == "Microsoft.VisualStudio.Code.Engine") | .value)
        } |
        select(.engine | ltrimstr("^") | split(".") |
            map(split("-")[0] | tonumber?) as $engine_parts |
            ($vscode_version | split(".") | map(tonumber)) as $vscode_parts |
            (
                ($engine_parts[0] // 0) < $vscode_parts[0] or
                (($engine_parts[0] // 0) == $vscode_parts[0] and ($engine_parts[1] // 0) < $vscode_parts[1]) or
                (($engine_parts[0] // 0) == $vscode_parts[0] and ($engine_parts[1] // 0) == $vscode_parts[1] and ($engine_parts[2] // 0) <= $vscode_parts[2])
            )
        ) |
        .version' | head -n 1
}

# Install extension
install_extension() {
    local extension_id="$1"
    local version="$2"
    local user_data_dir="$3"
    local extension_name
    extension_name=$(echo "$extension_id" | cut -d'.' -f2)
    local temp_dir="/tmp/code-extensions"

    echo "Installing $extension_id v$version..."

    # Create temp directory
    mkdir -p "$temp_dir"

    # Download
    echo "  Downloading..."
    # Use curl with portable options (--progress-bar not available everywhere)
    curl -L "https://marketplace.visualstudio.com/_apis/public/gallery/publishers/GitHub/vsextensions/$extension_name/$version/vspackage" \
        -o "$temp_dir/$extension_name.vsix.gz"

    if [ ! -f "$temp_dir/$extension_name.vsix.gz" ]; then
        echo "  ✗ Download failed for $extension_id"
        return 1
    fi

    # Decompress (handle both gunzip and gzip -d)
    if command -v gunzip >/dev/null 2>&1; then
        gunzip -f "$temp_dir/$extension_name.vsix.gz"
    else
        gzip -df "$temp_dir/$extension_name.vsix.gz"
    fi

    # Install with user-data-dir if provided
    if [ -n "$user_data_dir" ]; then
        code-server --user-data-dir="$user_data_dir" --force --install-extension "$temp_dir/$extension_name.vsix"
    else
        code-server --force --install-extension "$temp_dir/$extension_name.vsix"
    fi

    # Clean up
    rm -f "$temp_dir/$extension_name.vsix"

    echo "  ✓ $extension_id installed successfully!"
    return 0
}

# Check for required dependencies
check_dependencies() {
    local missing_deps=()

    # Check for required commands
    for cmd in curl jq code-server; do
        if ! command -v "$cmd" >/dev/null 2>&1; then
            missing_deps+=("$cmd")
        fi
    done

    # Check for either gunzip or gzip
    if ! command -v gunzip >/dev/null 2>&1 && ! command -v gzip >/dev/null 2>&1; then
        missing_deps+=("gunzip/gzip")
    fi

    if [ "${#missing_deps[@]}" -gt 0 ]; then
        echo "Error: Missing required dependencies: ${missing_deps[*]}"
        echo "Please install the missing dependencies and try again."
        exit 1
    fi
}

# Main script
echo "GitHub Copilot Extensions Installer"
echo "===================================="
echo ""

# Check dependencies
check_dependencies

# Get VS Code version
VSCODE_VERSION="$(get_vscode_version)"

if [ -z "$VSCODE_VERSION" ]; then
    echo "Error: Could not extract VS Code version from code-server"
    exit 1
fi

echo "Detected VS Code version: $VSCODE_VERSION"

# Check for user-data-dir in running code-server
USER_DATA_DIR="$(get_user_data_dir)"
if [ -n "$USER_DATA_DIR" ]; then
    echo "Detected user-data-dir: $USER_DATA_DIR"
fi
echo ""

# Extensions to install
# Use portable array declaration
EXTENSIONS="GitHub.copilot GitHub.copilot-chat"
FAILED=0

# Iterate through space-separated list for portability
for ext in $EXTENSIONS; do
    echo "Processing $ext..."

    # Find compatible version
    version="$(find_compatible_version "$ext" "$VSCODE_VERSION")"

    if [ -z "$version" ]; then
        echo "  ✗ No compatible version found for $ext"
        FAILED="$((FAILED + 1))"
    else
        echo "  Found compatible version: $version"
        if ! install_extension "$ext" "$version" "$USER_DATA_DIR"; then
            FAILED="$((FAILED + 1))"
        fi
    fi
    echo ""
done

# Summary
echo "===================================="
if [ $FAILED -eq 0 ]; then
    echo "✓ All extensions installed successfully!"
    # Clean up temp directory on success
    rm -rf /tmp/code-extensions
else
    echo "⚠ Completed with $FAILED error(s)"
    exit 1
fi
```

**Reply by Catmas06** [reply]
very thanks, works for me

**Reply by svefro** [reply]
> I created this https://github.com/sunpix/howto-install-copilot-in-code-server
> 
> installs latest compatible:
> 
> * copilot
> * copilot-chat

Thank You ! 😀

**Reply by oowl** [reply]
When i execute your script, error was throw,
```
GitHub Copilot Extensions Installer
====================================

Detected VS Code version: 4.108.2 3c0b449c6e6e37b44a8a7938c0d8a3049926a64c with Code 1.108.2

Processing GitHub.copilot...
jq: error (at <stdin>:1): Invalid numeric literal at line 1, column 43 (while parsing '2 3c0b449c6e6e37b44a8a7938c0d8a3049926a64c with Code 1')
  ✗ No compatible version found for GitHub.copilot

Processing GitHub.copilot-chat...
jq: error (at <stdin>:1): Invalid numeric literal at line 1, column 43 (while parsing '2 3c0b449c6e6e37b44a8a7938c0d8a3049926a64c with Code 1')
  ✗ No compatible version found for GitHub.copilot-chat

====================================
```

Now I change your script, works fine in my side.
```
# Extract VS Code version from code-server
get_vscode_version() {
    local version_line
    local code_version
    version_line="$(code-server --version 2>/dev/null | head -n 1)"
    if [ -z "$version_line" ]; then
        return 0
    fi

    # Extract "Code X.Y.Z" if present; otherwise fall back to first token.
    code_version="$(echo "$version_line" | sed -n 's/.*Code \([0-9][0-9.]*\).*/\1/p' | head -n 1)"
    if [ -n "$code_version" ]; then
        echo "$code_version"
    else
        echo "$version_line" | awk '{print $1}'
    fi
}

# Get user-data-dir from running code-server process
get_user_data_dir() {
    # Use ps with POSIX-compliant options
    local process_info
    if command -v ps >/dev/null 2>&1; then
        # Try BSD-style first (macOS), fallback to POSIX
        process_info=$(ps aux 2>/dev/null | grep -v grep | grep "code-server" | head -n 1) ||
        process_info=$(ps -ef 2>/dev/null | grep -v grep | grep "code-server" | head -n 1)
    fi

    if [ -n "$process_info" ]; then
        echo "$process_info" | grep -o -- '--user-data-dir=[^ ]*' | sed 's/--user-data-dir=//'
    fi
}

# Find compatible extension version
find_compatible_version() {
    local extension_id="$1"
    local vscode_version="$2"

    local response
    response=$(curl -s -X POST "https://marketplace.visualstudio.com/_apis/public/gallery/extensionquery" \
        -H "Content-Type: application/json" \
        -H "Accept: application/json;api-version=3.0-preview.1" \
        -d "{
            \"filters\": [{
                \"criteria\": [
                    {\"filterType\": 7, \"value\": \"$extension_id\"},
                    {\"filterType\": 12, \"value\": \"4096\"}
                ],
                \"pageSize\": 50
            }],
            \"flags\": 4112
        }")

    echo "$response" | jq -r --arg vscode_version "$vscode_version" '
        .results[0].extensions[0].versions[] |
        select(.version | test("^[0-9]+\\.[0-9]+\\.[0-9]*$")) |
        select(.version | length < 8) |
        {
            version: .version,
            engine: (.properties[] | select(.key == "Microsoft.VisualStudio.Code.Engine") | .value)
        } |
        select(.engine | ltrimstr("^") | split(".") |
            map(split("-")[0] | tonumber?) as $engine_parts |
            ($vscode_version | split(".") | map(tonumber)) as $vscode_parts |
            (
                ($engine_parts[0] // 0) < $vscode_parts[0] or
                (($engine_parts[0] // 0) == $vscode_parts[0] and ($engine_parts[1] // 0) < $vscode_parts[1]) or
                (($engine_parts[0] // 0) == $vscode_parts[0] and ($engine_parts[1] // 0) == $vscode_parts[1] and ($engine_parts[2] // 0) <= $vscode_parts[2])
            )
        ) |
        .version' | head -n 1
}

# Install extension
install_extension() {
    local extension_id="$1"
    local version="$2"
    local user_data_dir="$3"
    local extension_name
    extension_name=$(echo "$extension_id" | cut -d'.' -f2)
    local temp_dir="/tmp/code-extensions"

    echo "Installing $extension_id v$version..."

    # Create temp directory
    mkdir -p "$temp_dir"

    # Download
    echo "  Downloading..."
    # Use curl with portable options (--progress-bar not available everywhere)
    curl -L "https://marketplace.visualstudio.com/_apis/public/gallery/publishers/GitHub/vsextensions/$extension_name/$version/vspackage" \
        -o "$temp_dir/$extension_name.vsix.gz"

    if [ ! -f "$temp_dir/$extension_name.vsix.gz" ]; then
        echo "  ✗ Download failed for $extension_id"
        return 1
    fi

    # Decompress (handle both gunzip and gzip -d)
    if command -v gunzip >/dev/null 2>&1; then
        gunzip -f "$temp_dir/$extension_name.vsix.gz"
    else
        gzip -df "$temp_dir/$extension_name.vsix.gz"
    fi

    # Install with user-data-dir if provided
    if [ -n "$user_data_dir" ]; then
        code-server --user-data-dir="$user_data_dir" --force --install-extension "$temp_dir/$extension_name.vsix"
    else
        code-server --force --install-extension "$temp_dir/$extension_name.vsix"
    fi

    # Clean up
    rm -f "$temp_dir/$extension_name.vsix"

    echo "  ✓ $extension_id installed successfully!"
    return 0
}

# Check for required dependencies
check_dependencies() {
    local missing_deps=()

    # Check for required commands
    for cmd in curl jq code-server; do
        if ! command -v "$cmd" >/dev/null 2>&1; then
            missing_deps+=("$cmd")
        fi
    done

    # Check for either gunzip or gzip
    if ! command -v gunzip >/dev/null 2>&1 && ! command -v gzip >/dev/null 2>&1; then
        missing_deps+=("gunzip/gzip")
    fi

    if [ "${#missing_deps[@]}" -gt 0 ]; then
        echo "Error: Missing required dependencies: ${missing_deps[*]}"
        echo "Please install the missing dependencies and try again."
        exit 1
    fi
}

# Main script
echo "GitHub Copilot Extensions Installer"
echo "===================================="
echo ""

# Check dependencies
check_dependencies

# Get VS Code version
VSCODE_VERSION="$(get_vscode_version)"

if [ -z "$VSCODE_VERSION" ]; then
    echo "Error: Could not extract VS Code version from code-server"
    exit 1
fi

echo "Detected VS Code version: $VSCODE_VERSION"

# Check for user-data-dir in running code-server
USER_DATA_DIR="$(get_user_data_dir)"
if [ -n "$USER_DATA_DIR" ]; then
    echo "Detected user-data-dir: $USER_DATA_DIR"
fi
echo ""

# Extensions to install
# Use portable array declaration
EXTENSIONS="GitHub.copilot GitHub.copilot-chat"
FAILED=0

# Iterate through space-separated list for portability
for ext in $EXTENSIONS; do
    echo "Processing $ext..."

    # Find compatible version
    version="$(find_compatible_version "$ext" "$VSCODE_VERSION")"

    if [ -z "$version" ]; then
        echo "  ✗ No compatible version found for $ext"
        FAILED="$((FAILED + 1))"
    else
        echo "  Found compatible version: $version"
        if ! install_extension "$ext" "$version" "$USER_DATA_DIR"; then
            FAILED="$((FAILED + 1))"
        fi
    fi
    echo ""
done

# Summary
echo "===================================="
if [ $FAILED -eq 0 ]; then
    echo "✓ All extensions installed successfully!"
    # Clean up temp directory on success
    rm -rf /tmp/code-extensions
else
    echo "⚠ Completed with $FAILED error(s)"
    exit 1
fi

```

**Comment by oxFilla** [comment]
The only thing that doesn't work for me is the absence of the GitHub status icon in the lower status bar, which normally displays the remaining free usage and other informations.

**Reply by chris-demattio-keysight** [reply]
I see something similar.  I don't see the Copilot icon next to the command palette.  Everything else seems to be working, though.
